`fastrow` and `fastcolumns` documention and the defaults for memory layout stated clearly

arnavk23 opened the pull request:
If you want to submit an unfinished piece of work in order to get comments and discuss, please mark the pull request as a draft and ping the repository maintainer.

**Please address only one topic or issue per pull request! Many small PRs are much easier to review and merge than one large PR.**

Before merging, all changes and new functionality should be marked in the CHANGELOG file, but feel free to just leave your CHANGELOG notes in the PR description, to avoid merge conflicts with other requests modifying that file. The maintainer will add these CHANGELOG notes for you if you do so.

Before considering your pull request ready for review and merging make sure that all of the following are completed (please keep the clecklist as part of your PR):

- [ ] The code is properly formatted and commented.
- [ ] Substantial new functionality is documented within the docs.
- [ ] All new functionality is tested.
- [ ] All of the automated tests on github pass.
- [ ] We recently started enforcing formatting checks. If formatting issues are reported in the new code you have written, please correct them. <small>There will be plenty of old code that is flagged as we are slowly transitioning to enforced formatting. Please do not worry about or address older formatting issues -- keep your PR just focused on your planned contribution.</small>
- [x] Fixes #474


If possible, keep your git history not too wild (rebase and squash commits, keep commits small and semantically separated) so that review is easier.

arnavk23 committed:
reinterpretation

arnavk23 committed:
documentation

arnavk23 committed:
rechanging the implementation

arnavk23 committed:
making changes

arnavk23 committed:
Update Project.toml

arnavk23 committed:
Merge branch 'master' into master

arnavk23 committed:
Import Base.reinterpret for improved functionality

arnavk23 committed:
h

arnavk23 committed:
reimplementing tests

arnavk23 committed:
changes acc. to review

arnavk23 committed:
running JuliaFormatter

arnavk23 committed:
Merge branch 'master' into master

arnavk23 committed:
Update test/test_reinterpret.jl

Co-authored-by: Stefan Krastanov <github.acc@krastanov.org>

arnavk23 committed:
adding review changes

arnavk23 committed:
Merge branch 'master' into master

arnavk23 committed:
changes

arnavk23 committed:
reinterpret

arnavk23 committed:
reinterpret.jl

arnavk23 committed:
test

arnavk23 committed:
documentation for fastrow and fastcolumn data structures

arnavk23 committed:
Delete src/reinterpret.jl

arnavk23 committed:
Delete test/test_reinterpret.jl

arnavk23 committed:
Update QuantumClifford.jl

arnavk23 committed:
Merge branch 'master' into issue/474-fastrow-fastcolumn-docs

arnavk23 committed:
Merge branch 'master' into issue/474-fastrow-fastcolumn-docs

arnavk23 committed:
Merge branch 'master' into issue/474-fastrow-fastcolumn-docs

arnavk23 committed:
Merge branch 'master' into issue/474-fastrow-fastcolumn-docs

arnavk23 commented:
@Hamiltonian-Action can you review this pr too. Thanks!

Hamiltonian-Action commented on docs/src/datastructures.md:
```diff
@@ -71,7 +71,32 @@ Notice the results when the projection operator commutes with the state but is n
+
+The alternative **column-major (fastcolumn) layout** stores tableau columns (mostly) contiguously in memory. This layout is optimized for:
+- **Applying sparse gates** like `apply!(s, sCNOT(i,j))` - row updates on a few qubits
+- **Pauli multiplications** (left or right)
```
Do we actually have the data for this? From the GPU side, I would reckon that both styles should execute nearly equally as fast if implemented properly (assuming no phase information is desired, otherwise row-major would likely reign supreme)
Moreover, row-major multiplication is more straightforward to parallelise/vectorise/distribute/etc. as it eliminates any inter-task dependency.

Hamiltonian-Action commented on test/test_bitpack.jl:
```diff
@@ -80,4 +80,35 @@
         end
     end
+
+    @testset "memory layout performance comparison" begin
```
There is no performance comparison in these tests, they just serve to further verify that both layouts ought to produce indistinguishable results.
Performance comparison could/would/should take the form of some properly configured autonomous test script (or collection thereof) that generates benchmark results suitable for automatic inclusion in the documentation.
Stefan can correct me if I am wrong but I believe that is what the `benchmark` directory is for -- at least that is how I have utilised it in the past.

arnavk23 commented:
> Thanks, @arnavk23 , the enthusiasm you are showing in these PR is very heartening!

Thank you @Krastanov for your reply, I will surely keep these suggestions in mind working further on this repo and any other repo I ever work on. I am really enjoying going through the work done here by you and your lab and would love to contribute more. Thank you for everything.

arnavk23 committed:
clarification: Update FastRow and FastColumn documentation

Krastanov commented on docs/src/datastructures.md:
```diff
@@ -71,7 +71,36 @@ Notice the results when the projection operator commutes with the state but is n
+
+The alternative **column-major (fastcolumn) layout** stores tableau columns (mostly) contiguously in memory. This layout is optimized for:
+- **Applying sparse gates** like `apply!(s, sCNOT(i,j))` - row updates on a few qubits
+- **Pauli multiplications** (left or right)
```
To address HA's comments, you can  `@benchmark` examples that run during doc building.

E.g. `@benchmark canonicalize!(s) setup=(s=fastrow(random_stabilizer(200))) evals=1 samples=1`

Including something like this in the docs themselves would take care of most questions about data.

Krastanov reviewed (changes requested):
Thanks, this is on the way of becoming a really nice informative page!

You picked a good set of low-level functions to gain understanding of the codebase -- reinterpret/fastrow/fastcolumn can be subtle.

I left a few comments in. Please make sure to address HA's comments as well.

Feel free to convert to a non-draft when it is ready for review.

arnavk23 committed:
Documentation improvements for fastrow/fastcolumn layout functions

- Clarified memory layout concepts in user documentation
- Removed vague implementation details and outdated benchmark references
- Added concrete @time examples showing canonicalization performance differences
- Removed unfounded GPU performance claims without supporting data
- Renamed test to 'memory layout correctness' to reflect its actual purpose
- Documented that row-major layout shows measurable advantage on CPU for large states
- Added guidance to use profiling for layout-specific optimization decisions

arnavk23 committed:
Merge branch 'master' into issue/474-fastrow-fastcolumn-docs

arnavk23 committed:
Clarify fastrow/fastcolumn documentation and state defaults

- Explicitly document that fastrow is the default for Stabilizer, Destabilizer, MixedStabilizer, and MixedDestabilizer
- Explicitly document that fastcolumn is the default for PauliFrame
- Explain technical implementation: fastrow uses column-major storage, fastcolumn uses row-major storage
- Add indexing convention that applies regardless of layout: xzs[tableau_column_index, tableau_row_index]
- List which operations are optimal for each layout
- Update docs to use 'fastrow vs fastcolumn' terminology instead of 'row-major vs column-major'
- Add technical details section explaining how storage order changes between layouts

Addresses #635

arnavk23 committed:
Merge branch 'master' into issue/474-fastrow-fastcolumn-docs

Krastanov committed:
Merge branch 'master' into issue/474-fastrow-fastcolumn-docs

arnavk23 committed:
Merge branch 'QuantumSavory:master' into issue/474-fastrow-fastcolumn-docs

Hamiltonian-Action commented on docs/src/datastructures.md:
```diff
@@ -79,7 +79,76 @@ For circuits containing non-Clifford gates, two state types are available:
+- The **X component** of qubit `i` in stabilizer `j` is at `xzs[i_big, j]`  
+- The **Z component** is at `xzs[i_big + end÷2, j]`
+
+where `i_big` accounts for bit packing (multiple qubits packed into each `UInt64`).
```
`UInt64` is the default setting but not the only one. Explicitly stating this can lead to confusion when utilising other types.

Hamiltonian-Action commented on docs/src/datastructures.md:
````diff
@@ -79,7 +79,76 @@ For circuits containing non-Clifford gates, two state types are available:
+
+The performance difference becomes apparent when examining different operation types:
+
+```jldoctest
````
I would advise to avoid performing documentation tests against timing macros. This is prone to break from a plethora of different directions unless you have extensively tested the spread around these values to be absolutely miniscule.

Hamiltonian-Action commented on src/fastmemlayout.jl:
```diff
@@ -1,17 +1,48 @@
+This corresponds to column-major storage of the underlying `xzs` matrix.
 
-See also: [`fastrow`](@ref)"""
+**This is the default layout** for [`Stabilizer`](@ref), [`Destabilizer`](@ref), 
```
Pedantic, possibly out of scope, but perhaps it is better to avoid explicitly enumerating all the various structures and merely say that this is the default layout unless explicitly stated otherwise?

Hamiltonian-Action commented on test/test_bitpack.jl:
```diff
@@ -80,4 +80,33 @@
         end
     end
+
+    @testset "memory layout correctness" begin
```
The `copy` invocations inside `canonicalize!` and `apply!` are extraneous given how the inputs are constructed as `x = fastrow(random_stabilizer(...))` and `y = fastcolumn(copy(x))`.

Hamiltonian-Action reviewed (commented):
Pedantic remarks, shouldn't require any major modifications.

arnavk23 committed:
Update fastmemlayout.jl

arnavk23 committed:
Update test_bitpack.jl

arnavk23 committed:
Update datastructures.md

Krastanov committed:
minor cleanup
